### PR TITLE
win-capture: Expose internal handles using dynamic procedures

### PR DESCRIPTION
**Pull request by henke37**
### Description
This adds dynamic procedures and a signal to the window and game capture sources in order to expose internal handles in a stable way.

### Motivation and Context
This addition lets 3rd parties access the window handle to the captured window without having to dig into a bunch of implementation internal structs.

### How Has This Been Tested?
Loaded up the tweaked win-capture plugin in my own build of OBS and verified no crashes and no errors in the log.
Updated my window follower plugin to make use of this new system.
Loaded up my plugin along with the win-capture plugin in my own build of OBS and verified no crashes or errors in the log. Made sure to test that triggering a new window handle in win-capture worked correctly.

### Types of changes

- New feature (non-breaking change which adds functionality)
Exposes handles in a safe way. Adds a signal to let users know when the window handle has changed. Adds a new libobs function, `proc_handler_exists`, to check if a dynamic procedure exists without calling it.

### Checklist:

- [ ] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [X] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [X] My code is not on the master branch.
- [X] The code has been tested.
- [X] All commit messages are properly formatted and commits squashed where appropriate.

### CODE REVIEW

File: libobs/callback/proc.c [hunk 1]
```diff
@@ -90,3 +90,19 @@ bool proc_handler_call(proc_handler_t *handler, const char *name,
...
+
+bool proc_handler_exists(proc_handler_t *handler, const char *name)
+{
+	if (!handler)
+		return false;
+
+	for (size_t i = 0; i < handler->procs.num; i++) {
+		struct proc_info *info = handler->procs.array + i;
+
+		if (strcmp(info->func.name, name) == 0) {
```
Comment (RytoEX) [inline libobs/callback/proc.c:102]: Don't use curly braces in a single statement `if`.

**Comment by henke37** [comment]
I was under the impression that I did prefix each commit with the component that it modifies.

**Comment by RytoEX** [comment]
> I was under the impression that I did prefix each commit with the component that it modifies.

I'm pretty sure that the third commit was previously prefixed with "win-capture" instead of "libobs".  Also, the final commit is still prefixed "Docs" rather than "docs".

**Comment by henke37** [comment]
Nah, it was always prefixed as `lib-obs`. Anyway, this should be the last change.

File: docs/sphinx/reference-libobs-callback.rst [hunk 2]
```diff
@@ -286,3 +286,15 @@ direct access to declarations or callback pointers.
    :param handler: Procedure handler object
    :param name:    Name of procedure to call
    :param params:  Calldata structure to pass to the procedure
+   
+
+---------------------
+
+.. function:: bool proc_handler_exists(proc_handler_t *handler, const char *name, calldata_t *params)
```
Comment (RytoEX) [inline docs/sphinx/reference-libobs-callback.rst:293]: Unexpected parameter `params` in documentation.  The actual function `proc_handler_exists` only takes two parameters:  `proc_handler_t *handler` and `const char *name`.

**Review by RytoEX — CHANGES_REQUESTED** [review]
This PR complies with the project's style and commit guidelines as far as I can tell.  It will still need to be compiled and functionally tested.  After making the requested change, I can approve it _only_ for style and commit guideline compliance.

It might look better to potential reviewers if you confirm that this complies with the project's clang-format requirements by checking the appropriate box in the PR template.

File: docs/sphinx/reference-libobs-callback.rst (hunk 2 above)
Comment (henke37) [reply docs/sphinx/reference-libobs-callback.rst:293]: Stupid mistake. Fixed.

File: plugins/win-capture/window-capture.c [hunk 3]
```diff
@@ -257,6 +264,12 @@ static void *wc_create(obs_data_t *settings, obs_source_t *source)
 		}
 	}
 
+	proc_handler_t *ph = obs_source_get_proc_handler(source);
+	proc_handler_add(ph, "void getWindowHandle(out ptr window)",
```
Comment (RytoEX) [inline plugins/win-capture/window-capture.c:268]: Use snake case for the proc handler declaration string too.

File: docs/sphinx/reference-libobs-callback.rst [hunk 4]
```diff
@@ -286,3 +286,15 @@ direct access to declarations or callback pointers.
...
+---------------------
+
+.. function:: bool proc_handler_exists(proc_handler_t *handler, const char *name)
+
+   Checks if a function exists in a procedure handler.
+
+   :param handler: Procedure handler object
+   :param name:    Name of procedure to check for
+   
+
```
Comment (RytoEX) [inline docs/sphinx/reference-libobs-callback.rst:300]: Extra newline can be deleted.  See rest of file for spacing patterns.

File: libobs/callback/proc.h [hunk 5]
```diff
@@ -49,6 +49,9 @@ EXPORT void proc_handler_add(proc_handler_t *handler, const char *decl_string,
 EXPORT bool proc_handler_call(proc_handler_t *handler, const char *name,
 			      calldata_t *params);
 
+/** Checks if a function exists in a procedure handler. Returns true if the named procedure is found. */
```
Comment (RytoEX) [inline libobs/callback/proc.h:52]: Nitpicks:
* "Check" instead of "Checks"
* "Return" instead of "Returns"

**Review by RytoEX — CHANGES_REQUESTED** [review]
After these changes, I'll hopefully be done.  As previously stated, someone else will have take up the functional test/review of the code.

File: libobs/callback/proc.h (hunk 5 above)
Comment (henke37) [reply libobs/callback/proc.h:52]: This seems inconsistent with the comment above this one.

File: plugins/win-capture/game-capture.c [hunk 6]
```diff
@@ -538,6 +538,20 @@ static void game_capture_update(void *data, obs_data_t *settings)
 
 extern void wait_for_hook_initialization(void);
 
+static void game_capture_get_windowHandle(void *data, calldata_t *cd)
```
[the remark this answers is not in the document]
Comment (henke37) [reply plugins/win-capture/game-capture.c:541]: Fixed.

File: libobs/callback/proc.h (hunk 5 above)
Comment (RytoEX) [reply libobs/callback/proc.h:52]: Then leave it as is.

Comment (henke37) [reply libobs/callback/proc.h:52]: Righty. That should be the last change.

**Comment by EZ64cool** [comment]
Bit of a necro, but is there any chance of this being merged? Seems to have been left in a ready state then dropped off.
Would be super useful for my plugin 😃

**Comment by spin0za** [comment]
@RytoEX Sorry, is there a chance this gets approved? I see all change requests resolved.
This is the only plugin that allows tracking moving windows. I can't imagine using OBS without such a basic functionality.

**Comment by Fenrirthviti** [comment]
Just going to log for posterity the notion this is "basic functionality" is very optimistic, at best. Window capture already works without needing to know the position of a window, and we're still unsure what workflows this really allows that aren't already accomplished with other, less convoluted methods. This still needs to be reviewed to see if it's something we want to accept, but I flat reject the notion this is "basic functionality" when we can't even really figure out what problem plugins using this is trying to solve.

**Comment by RytoEX** [comment]
> Bit of a necro, but is there any chance of this being resolved? Seems to have been left in a ready state then dropped off and out of sync. Would be super useful for my plugin 😃

First and foremost, it would need to not have merge conflicts. Second, we would need a compelling motivation to accept these changes.

What plugin and why would it be useful? Why is this _necessary_ for us to add and maintain here? What wide-reaching benefits does it provide?

As [I've mentioned before](https://github.com/obsproject/obs-studio/pull/2659#pullrequestreview-611050091), someone else would have to do the functional test/review of the code. No one else has seemed inclined to do so.


> @RytoEX Sorry, is there a chance this gets approved? I see all change requests resolved. This is the only plugin that allows tracking moving windows. I can't imagine using OBS without such a basic functionality.

We aren't in the habit of adding new functionality that will see no use in OBS itself without demonstrating that it is _required_, or very nearly so, for some third-party functionality that has wide-reaching benefits. Essentially, we want to make sure that the additional maintenance burden of adopting this code is worth it for us to maintain.

Furthermore, I, at least, do not consider this "basic functionality", and I think there would be more outcry/demand for this feature if it was universally agreed that it was so.

In any case, as [previously stated by me](https://github.com/obsproject/obs-studio/pull/2659#pullrequestreview-611050091):
> As previously stated, someone else will have take up the functional test/review of the code.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Merge the changes (using git rebase and fixup) for the `docs` commit into the `libobs` commit that added the new functions being documented. That way, checking out that commit doesn't end up with an undocumented function. Keep the `libobs: ` commit message subject prefix.

This review is not tacit approval of this PR. I am just noting issues that need to be fixed.

**Comment by spin0za** [comment]
Thanks for your reply :) @RytoEX @Fenrirthviti 

> we're still unsure what workflows this really allows that aren't already accomplished with other, less convoluted methods
Sorry what other less convoluted methods are you referring to? I'd love to know. In case you don't know what this plugin solves, check out this video (timestamped): https://youtu.be/rlLccEhaCAc?si=PkBiwrdyarRyn78B&t=193

> I, at least, do not consider this "basic functionality", and I think there would be more outcry/demand for this feature if it was universally agreed that it was so
I guess most streamers already use an overlay with fixed window size/position so they might not need this. But for those who don't, like me, I don't want my window to be sticked to the upper left corner. That is just bad viewing experience. I would either have to maximize my window or manually drag it in OBS scenes. So I don't understand why it's not basic.

**Comment by Fenrirthviti** [comment]
That video is explaining what the plugin does, not the why. I understand full well *what* it does, but the *why* is no really obvious. Why do you need a window position to be tracked in realtime in your OBS scene? What workflow is that actually serving? Why is this worth us accepting a complex feature and significant technical debt to support? Those are the questions we're trying to answer here. I am genuinely not sure what kind of setup would need a window to follow the location around the screen where display capture wouldn't suffice, and so far nobody has been able to explain a compelling use case for this feature.

**Comment by EZ64cool** [comment]
Hey, in terms of my usecase in https://github.com/EZ64cool/obs-hadowplay , this would be really useful to be able to grab the captured window's product name for automatic replay organisation. At the moment I have to manually check for fullscreen windows whilst a game-capture is active, this then means my plugin can't fully function with game-captures that either capture specific windows or the foreground on shortcut.

I believe that exposing the handle is a nice way of doing this as it allows a large amount of flexability within different scenarios. e.g. getting the window position/grabbing the product name.

Also considering @henke37 hasn't responded in a while I'd imagine this would have to come back as a new FQ, unless somehow FQ can be transferred? I'm still learning github/git

**Comment by EZ64cool** [comment]
I've actually just come across [`get_hooked`](https://docs.obsproject.com/reference-sources?highlight=get_hooked) in the docs which should be enough for my usecase. And if I'm not looks like it does a very similar behaviour to this PR.

**Comment by henke37** [comment]
Life has kept me busy, and I haven't had anything to add to the conversation until now.

As for explaining the why, shenanigans. My use case for the partner plugin was not manually controlled window movements, but rather automatic window movements. The kind of effects that would be a pain to capture, either requiring full screen capture (non ideal for many situations) or a virtualized windowing environment (not always provided).

There is also the opportunity for plugins to directly access the captured window. Be it by sending window messages or using the process handle. To give a concrete example: a speedrun timer that accesses game state to automate splits and ignore load times.

And while I do consider this feature advanced, the implementation is not complex code. It is more boilerplate than action, it uses documented apis in obvious ways and does nothing tricky.

**Comment by spin0za** [comment]
> the _why_ is not really obvious. Why do you need a window position to be tracked in realtime in your OBS scene? What workflow is that actually serving?

I wonder if you read my last comment. But I will repeat here. Without this plugin, for a scene where you do a window capture only, the captured source would stick to a fixed location (default upper left). I will show a screenshot below, where I editted the transform so that there is some margin on top and left, for better viewing experience. But then if I maximize the app, the transform won't adapt, causing a part of my app window being trimmed (here you can't see the textbox in Discord).

So basically, this makes using apps with dynamic sizes very awkward. You can't expect users to manually change the transform every time they resize a window. Heck, even the "center to screen" option is adaptive as I would imagine: once you maximize the app half the content is trimmed.

Discord with manually entered margin:
Screenshot 2024-03-25 19-03-26

Discord maximized, no margin change:
Screenshot 2024-03-25 19-03-54

**Comment by PatTheMav** [comment]
I've tried to follow the discussion around what problem or user need this PR attempts to solve and I'm not convinced this is something that we should add:

* It exposes an _extremely_ internal implementation detail of the window capture that can (and might) change in the future depending on where Windows' APIs might go. The handle is not something that 3rd party code should expect or even be exposed to.
* I'm also not convinced of the approach to implement functionality that has zero use for OBS Studio itself and only exists to enable a very specific want of a plugin. 

That would be a hierarchy inversion between OBS Studio and plugins, because core functionality is implemented _for_ OBS Studio and plugins are allowed to tap into what is available, but not the other way around.

* The actual user need that I can see is being able to set up a scene that uses a window capture that keeps a certain visual appearance regardless of the size of the captured window.

That particular need should be possible to achieve using a bounding box that ensures that the window capture source stays where it is and the visual content never exceeds that box. Using the window handle and low-level APIs to somehow keep track of the actual window size to somehow have this feed back into the size of the source is very much doing "brain surgery via the knee".

So this gets a "no" from me. The handle is a very internal implementation detail of the window capture, something that nothing outside of the code of that plugin itself should even be aware of (and I would argue that `get_hooked` itself is already questionable as it's too specific, there should be a more generic way for 3rd party code to provide metadata about what content they provide which would include the name of a captured process or window for capture-type sources).

**Comment by henke37** [comment]
The Windows API is not going to stop using HWNDs, ever. They were there back in 1.0 and are still there today. They are how applications interact with windows.

As for user needs, this is an extension point. Plugins can use the handles to interact with the captured content. This is what is needed for things like an auto-splitter or game stats display.

**Comment by Fenrirthviti** [comment]
I had asked for a reasonable use case that would be worth us maintaining this new code, and in my opinion, the answer being "shenanigans" back then was reason enough to close this. We are not a playground project for experiments, and a broader use case for this outside a hyper-specific vanity showcase was never really provided.

This should be maintained in a fork.

**Comment by spin0za** [comment]
> the answer being "shenanigans" back then was reason enough to close this

What's your definition of shenanigans? I described my needs in comments above, and it's a pretty common scenario that you change your Discord's location/size between conversation and watching others stream. Now without this plugin, you can only track this change with Display Capture. But there is plenty reason not to use Display Capture, as you don't want to show random stuff on your screen while streaming.

**Comment by Fenrirthviti** [comment]
> > the answer being "shenanigans" back then was reason enough to close this
> 
> What's your definition of shenanigans? I described my needs in comments above, and it's a pretty common scenario that you change your Discord's location/size between conversation and watching others stream. Now without this plugin, you can only track this change with Display Capture. But there is plenty reason not to use Display Capture, as you don't want to show random stuff on your screen while streaming.

As described by Pat above, while this use case seems reasonable at the surface, using this method to implement that is doing brain surgery through the knee. I am not saying there aren't valid use cases that could be implemented with this, but I think it's important to understand that simply exposing internal handles to third-parties is extremely risky and heavy-handed for something simple like basic window position.

Additionally, your use case could be solved by simply using a bounding box on the source and having the source scale to the bounds so it doesn't adjust off-screen. You don't need detailed screen position information. Again, brain surgery, knee, etc.

**Comment by henke37** [comment]
Not a fan of this whole "brain surgery via the knee" analogy. I do not see it as being needlessly indirect. This is very direct. As for the "risk", I'd like to hear your risk assessment. Because mine clearly differs.

The current workaround of peeking at internal, undocumented, structures is by any means of comparison more risky than a clear documentable api where encapsulation is not violated. At least not to the flagrant extent of internal structures. It's even forward compatible, if there for some reason isn't a HWND to get then the code will either fail to find the procedure handler or safely get an invalid handle. Not read junk data.

As for the handles themselves, HWNDs aren't kernel handles and can't be closed or otherwise mismanaged. It'd be just as dangerous if the HWND was obtained independently.

I am also unsure what you mean by "heavy handed". There is no meaningful lighter option. This is an extension point, and due to its nature there is no feasible way to predict all possible use cases and provide "safer" access methods specialized for them.

**Comment by spin0za** [comment]
> Window capture already works without needing to know the position of a window

The question is why must it _not_ know the position of a window? It doesn't make sense to me why placing a window always at topleft is good.

> your use case could be solved by simply using a bounding box on the source and having the source scale to the bounds so it doesn't adjust off-screen

This requires tedious manual setup for every single source, which could be totally avoided. Besides, still no solution to track a moving window.

**Comment by Fenrirthviti** [comment]
Again, we understand the request here seems to boil down to "I want to track the coordinates of the window", and that is a valid request. However, there is likely a better way of achieving this than exposing the entire handle. This is risky, and too easy to abuse and isn't something we should be just blindly trusting to any third-party plugin as part of our internal API. It's bad practice, and becomes a maintenance nightmare as by exposing this, we are implicitly offering to support it.

**Comment by henke37** [comment]
To begin with, the window position is one of several use cases. We should not focus too much on it. This is an extension point, not a window position detail.

Again, risky how? Easy to "abuse" how?

You mention "trust". I fail to see how plugins aren't already trusted if they are loaded into the same process without even an attempt at sandboxing.

As for the "maintenance nightmare", this is not a lot of code, around 35 interesting lines, depending on how you count. It's to the point and doesn't do anything tricky.

From the perspective of locking down internals, not only is this not something likely to change since it is an external matter: being OS API dependent, and having a proven history of remaining stable between os versions, but I've already explained how plugin code would deal with major breaking changes here.

So far, all arguments against this have been baseless assertations. A lot of "this is X" and "this is Y" without any details, definitions, clarifications or explanations.
